MultiModel supports evolution

jajcayn opened the pull request:
MultiModels now work with evolution.

caglorithm reviewed (changes requested):
Could you move the star dict related code from evolution.py to parameterSpace.py or dotdict.py?

jajcayn commented on neurolib/optimize/evolution/evolution.py:
```diff
@@ -287,7 +279,16 @@ def getModelFromTraj(self, traj):
+            individual_params = {
+                key_u: v
+                for k, v in individual_params.items()
+                for key_u in list(self.model.params[k.replace("STAR", "*")].keys())
```
unfortunately, it must be there - the difference between `star_dotdict` and simple params needs to be addressed when the model is actually updated with new parameters and that is inside Evolution or Exploration..
the problem is with `namedtuple` (hence only Evolution thanks to `deap`) because you cannot have "*" inside strings that serve as the "name" in the name tuple... so I simply replace actual "*" with the word "STAR" and then I change it back... I can move the functions for this to `collections...`

jajcayn commented on neurolib/utils/parameterSpace.py:
```diff
@@ -49,9 +49,12 @@ def __init__(self, parameters, parameterValues=None, kind=None, allow_star_notat
-            self.named_tuple_constructor = namedtuple("ParameterSpace", parameters)
-            self.named_tuple = self.named_tuple_constructor(*self.parameterValues)
+        # namedtuple names must be identifiers, so swap "*" for string "STAR"
+        def _unstar_dict(dct):
+            return {k.replace("*", "STAR"): v for k, v in dct.items()}
+
+        self.named_tuple_constructor = namedtuple("ParameterSpace", _unstar_dict(parameters))
+        self.named_tuple = self.named_tuple_constructor(*self.parameterValues)
```
ok, I guess I can move it

jajcayn commented:
moved all `star_dotdict` functions necessary for evolution and exploration to run to `utils.collections` and also added test for them... need to keep coverage as high as possible:)
